rust/src/derivation.rs

```diff
@@ -579,6 +540,175 @@ pub unsafe extern "C" fn zcashlc_get_sapling_receiver_for_unified_address(
     unwrap_exc_or_null(res)
 }
 
+/// Constructs an ffi::AccountMetadataKey from its parts.
+///
+/// # Safety
+///
+/// - `sk` must be non-null and valid for reads for 32 bytes, and it must have an alignment of `1`.
+/// - `chain_code` must be non-null and valid for reads for 32 bytes, and it must have an alignment
+///   of `1`.
```

Comment:
```suggestion
/// - `sk` must be non-null and valid for reads for 32 bytes, and it must have an alignment of `1`.
/// - The memory referenced by `sk` must not be mutated for the duration of the function call.
/// - `chain_code` must be non-null and valid for reads for 32 bytes, and it must have an alignment
///   of `1`.
/// - The memory referenced by `chain_code` must not be mutated for the duration of the function
///   call.
```

Reply:
There might be other missing cases of "The memory referenced by `...` must not be mutated for the duration of the function call." I didn't check.

Reply:
Technically we don't need it to be unmutated for the duration of the function call, as we read the pointer as a slice and then immediately convert into a stack-allocated array (thus no longer depending on the original pointer's memory). Compare this to `seed` where we read the pointer as a slice, and then pass that slice into other functions.

That being said, conveying a requirement that the memory not be modified for "the first few operations within the function call" is hard, so I'm fine with having a much stricter requirement than we actually need.